### PR TITLE
For fix slow dependency issue(base on 1.0.0).

### PR DESCRIPTION
**Pull request by yanzhiwei147**
for issue #5180

### CODE REVIEW

**Comment by neonichu** [comment]
Hey @yanzhiwei147 please fix the rubocop errors: https://travis-ci.org/CocoaPods/CocoaPods/jobs/140974410#L3353-L3360 — you can see them locally by running `bundle exec rake rubocop`

**Comment by yanzhiwei147** [comment]
@neonichu It's OK now!

**Comment by jcampbell05** [comment]
Do we have a benchmark test for this ? Would be useful to have something we can mesure.

**Comment by neonichu** [comment]
Yah, there was a small benchmark in <https://github.com/CocoaPods/CocoaPods/pull/5510#issuecomment-225811139>

@yanzhiwei147 Thanks for updating the PR, could you please rebase it one last time to get rid of the merge commits — we'd like to not have those in the history. Alternatively, you can also squash if you like.

**Comment by yanzhiwei147** [comment]
@neonichu So how can i do it for your mean? What's the history?

**Comment by neonichu** [comment]
@yanzhiwei147 Sorry, I was talking about the Git commit history. The one for your branch has a couple of merge commits, which you could remove by rebasing the branch. Another option is for you to squash the whole Git history of the branch into a single commit.

File: lib/cocoapods/resolver.rb [hunk 1]
```diff
@@ -73,9 +73,10 @@ def resolve
     def specs_by_target
       @specs_by_target ||= {}.tap do |specs_by_target|
         podfile.target_definition_list.each do |target|
+          cached_dependencys = {}
```
Comment (segiddins) [inline lib/cocoapods/resolver.rb]: `dependencies`

File: lib/cocoapods/resolver.rb [hunk 2]
```diff
@@ -471,13 +472,25 @@ def spec_is_platform_compatible?(dependency_graph, dependency, spec)
     #         An array of target-appropriate nodes whose `payload`s are
     #         dependencies for `target`.
     #
-    def valid_dependencies_for_target_from_node(target, node)
+    def valid_dependencies_for_target_from_node(target, cached_dependencys, node)
+      cache_dependency = cached_dependencys[node.name]
```
Comment (segiddins) [inline lib/cocoapods/resolver.rb]: `cached_dependencies ||= begin`, that way all the code that does the calculation can be inside the `begin..end` block

File: CHANGELOG.md [hunk 3]
```diff
@@ -23,6 +23,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 * [repo/push] --use-json to convert podspecs to JSON format when pushing  
   [Mark Schall](https://github.com/maschall)
   [#5568](https://github.com/CocoaPods/CocoaPods/pull/5568)
+  
+* Improve the performance of resolve dependency.
```
Comment (segiddins) [inline CHANGELOG.md]: `Improve performance of dependency resolution.`

**Comment by segiddins** [comment]
@yanzhiwei147 if you're interested in making those changes, we can get this merged :)

**Comment by yanzhiwei147** [comment]
@segiddins OK.

File: lib/cocoapods/resolver.rb (hunk 2 above)
Comment (yanzhiwei147) [inline lib/cocoapods/resolver.rb]: Sorry, I can't understand this comment.

**Comment by segiddins** [comment]
My comments from before still hold -- if you want, someone on the team can try and take this PR over, although there's no guarantee when we'll be able to get around to it

**Comment by yanzhiwei147** [comment]
I fix most of your comments  suggest.

File: lib/cocoapods/resolver.rb [hunk 4]
```diff
@@ -471,13 +472,25 @@ def spec_is_platform_compatible?(dependency_graph, dependency, spec)
     #         An array of target-appropriate nodes whose `payload`s are
     #         dependencies for `target`.
     #
-    def valid_dependencies_for_target_from_node(target, node)
+    def valid_dependencies_for_target_from_node(target, dependencies, node)
+      cache_dependency = dependencies[node.name]
```
Comment (segiddins) [inline lib/cocoapods/resolver.rb]: this should be using `dependencies[node.name] ||=`

**Comment by yanzhiwei147** [comment]
@segiddins It's OK now!

**Comment by cheniuiOSloved** [comment]
1.1.0.beta.1 version has fixed the problem ? @yanzhiwei147

File: lib/cocoapods/resolver.rb [hunk 5]
```diff
@@ -471,13 +472,21 @@ def spec_is_platform_compatible?(dependency_graph, dependency, spec)
...
 
-      dependency_nodes + dependency_nodes.flat_map { |n| valid_dependencies_for_target_from_node(target, n) }
+        dependency_nodes + dependency_nodes.flat_map do |item|
+          node_result = valid_dependencies_for_target_from_node(target, dependencies, item)
+
+          node_result
+        end
+      end
+
+      dependencies[node.name]
```
Comment (segiddins) [inline lib/cocoapods/resolver.rb]: this last line is redundant

File: CHANGELOG.md [hunk 6]
```diff
@@ -67,6 +67,10 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
 * Add a license type to generated acknowledgements file in plist.  
   [Naoto Kaneko](https://github.com/naoty)
   [#5436](https://github.com/CocoaPods/CocoaPods/pull/5436)
+  
+* Improve performance of dependency resolution.
```
Comment (segiddins) [inline CHANGELOG.md]: this needs to be in the `master` section of the changelog

**Comment by yanzhiwei147** [comment]
@cheniuiOSloved YES, it can be fixed at 1.1.0.beta.

File: CHANGELOG.md (hunk 6 above)
Comment (yanzhiwei147) [inline CHANGELOG.md]: OK

File: lib/cocoapods/resolver.rb (hunk 5 above)
Comment (yanzhiwei147) [inline lib/cocoapods/resolver.rb]: OK

**Comment by yanzhiwei147** [comment]
@segiddins It should merged [Cocoapods Core](https://github.com/CocoaPods/Core/pull/340)

**Pull request merged**